src/components/DragBox/DragBox.js

```diff
@@ -0,0 +1,271 @@
+      x: Math.min(start.x, end.x) - window.scrollX,
+      y: Math.min(start.y, end.y) - window.scrollY,
+      width: Math.abs(start.x - end.x),
+      height: Math.abs(start.y - end.y),
+    };
+  };
+
+  // Checks the collisions with the draggable box against the children. The collision deltas are then used
+  // to invoke callbacks that consumers can bind to.
+  checkChildrenBoxCollisions = () => {
+    const {
+      props: { onCollisionBegin, onCollisionEnd, onCollisionChange },
+      state: { collisions },
+    } = this;
+    const collisionBox = this.calcRect();
+    // Divide all clickable refs into two arrays based on whether they are colliding or not.
+    const [colliding, notColliding] = partition(
+      Object.keys(this.childElements),
+      key => {
+        const node = ReactDOM.findDOMNode(this.childElements[key]);
+        const childRect = node.getBoundingClientRect();
+        return this.checkBoxCollision(collisionBox, childRect);
+      },
+    );
+
+    // Pull out elements that haven't changed their collision state.
+    const added = colliding.filter(key => !collisions.has(key));
+    const removed = notColliding.filter(key => collisions.has(key));
+
+    //Update collisions set
+    added.forEach(i => collisions.add(i));
+    removed.forEach(i => collisions.delete(i));
+
+    //Invoke callbacks if necessary
+    if (added.length > 0) {
+      onCollisionBegin(added);
+    }
+    if (removed.length > 0) {
+      onCollisionEnd(removed);
+    }
+    if (added.length > 0 || removed.length > 0) {
+      onCollisionChange({ added, removed });
+    }
+    this.setState({ collisions });
+  };
+
+  checkBoxCollision = (b1, b2) =>
+    b1.x < b2.x + b2.width &&
+    b1.x + b1.width > b2.x &&
+    b1.y < b2.y + b2.height &&
+    b1.height + b1.y > b2.y;
+
+  shouldDrawRect = rect => {
+    const { state: { mouseDown } } = this;
+    return mouseDown && rect.width > 0 && rect.height > 0;
+  };
+
+  renderRect = () => {
+    const {
+      props: { renderRect },
+      state: { mouseDown },
+      calcRect,
+      shouldDrawRect,
+    } = this;
+    const rect = calcRect();
+    if (!shouldDrawRect(rect)) return;
+    return renderRect({
+      left: rect.x,
+      top: rect.y,
+      width: rect.width,
+      height: rect.height,
+    });
+  };
+
+  getRef = key => el => (this.childElements[key] = el);
+
+  renderContents = () => {
+    const {
+      props: {
+        onCollisionChange,
+        onCollisionBegin,
+        onCollisionEnd,
+        onMouseUp,
+        onMouseDown,
+        renderContents,
+      },
+      state: { collisions },
+      getRef,
+    } = this;
+    return renderContents({
+      onMouseDown,
+      onMouseUp,
+      onCollisionChange,
+      onCollisionBegin,
+      onCollisionEnd,
+      collisions,
+      getRef,
+    });
+  };
+
+  render() {
+    const { onMouseDown, renderRect, renderContents } = this;
+    return (
+      <div ref={el => (this.dragElement = el)} onMouseDown={onMouseDown}>
```

Comment:
ref could be moved to class function, or you could use `React.createRef` to do refs the recommended 'clean' way (I just learned about this today) https://reactjs.org/docs/refs-and-the-dom.html